narwhal/primary/src/core.rs

```diff
@@ -809,7 +652,54 @@ impl Core {
                 },
 
                 // We also receive here our new headers created by the `Proposer`.
-                Some(header) = self.rx_headers.recv() => self.process_own_header(header).await,
+                Some(header) = self.rx_headers.recv() => {
+                    let (tx_cancel, rx_cancel) = oneshot::channel();
+                    if let Some(cancel) = self.cancel_proposed_header {
+                        let _ = cancel.send(());
+                    }
+                    self.cancel_proposed_header = Some(tx_cancel);
+
+                    self.try_update_committee().await;
+
+                    let name = self.name.clone();
+                    let committee = self.committee.clone();
+                    let header_store = self.header_store.clone();
+                    let certificate_store = self.certificate_store.clone();
+                    let signature_service = self.signature_service.clone();
+                    let metrics = self.metrics.clone();
+                    let network = self.network.network();
+                    self.propose_header_future = Some(spawn_monitored_task!(Self::propose_header(
+                        name,
+                        committee,
+                        header_store,
+                        certificate_store,
+                        signature_service,
+                        metrics,
+                        network,
+                        header,
+                        rx_cancel,
+                    ))).into();
```

Comment:
What's the expected behavior if there was already an outstanding propose_header_future?

Reply:
It should abandon the old header proposal and start working on the new one instead. Do you have any concerns about the way it's implemented?

Reply:
No concerns, just wanted to check what the expectation was

Reply:
seems like we should make sure we are proposing one header at a time

Reply:
I suppose we leverage the cancel handle in order to cancel the other running task. Given we have direct access to the join handle i wonder if it would be easier to handle this by explicitly aborting the task. (Today this works just fine so i don't expect you to change this)

Reply:
I considered and then explicitly decided not to abort the task because I don't want the code to stop executing in between await points and have it potentially mess up DB store writes, metric updates, etc. So I put in an explicit cancel that is honored at a safe point.